Make client requests by URI

djmetzle opened the pull request:
Make the uri object for endpoints first class.

Add a class `ApiUri` that handles putting together the full endpoint
urls. This includes GET params.

Modify the client class to accept these full uri objects when making
requests.

djmetzle committed:
Add a class to build a URI object

This uses the standard format for the Replicated Vendor API to build up
a uri object for an endpoint. With this simplified interface, the
request client can be simplified, and the API modules only need to be
concerned with building the correct `endpoint` string.

djmetzle committed:
Use URI for client request

Change the request interface to take a full `URI` object instead of a
string path.

This simplifies the structure of the request method, and eliminates the
`encode_path_params` method completely.

djmetzle committed:
Update the ctor

We need to create a persistent https connection to the API.

g1y commented on lib/client/client.rb:
```diff
@@ -4,40 +4,37 @@
 
 class ApiClient
-   ENDPOINT = "https://api.replicated.com/vendor/v1"
+   VENDOR_ENDPOINT = "https://api.replicated.com"
```
vendor/v1?

g1y replied:
nvm, we are only using the host and port. May be better to just change this to `VENDOR_HOST` and add a `VENDOR_PORT` const

djmetzle committed:
Make initializing connection more explicit

Use a hardcoded hostname to the api server, and the default HTTPS port
from the URI module to instantiate the connection.

The `URI.parse` format was unclear.

MooreGuy commented:
CR :+1:

djmetzle commented:
Mergin' it.
